### PR TITLE
output: prevent stage file from being used as output

### LINKED ISSUE #2345: run: warning/exception when dependency is a dvc-file
It doesn't make a lot of sense to depend on a dvc-file, however, the interface is not intuitive enough to prevent this.

If we want to allow this behavior, it should be with a warning.

### PR DESCRIPTION
**Pull request by pared**
* [x] Have you followed the guidelines in our
      [Contributing document](https://dvc.org/doc/user-guide/contributing)?

* [x] Does your PR affect documented changes or does it add new functionality
      that should be documented? If yes, have you created a PR for
      [dvc.org](https://github.com/iterative/dvc.org) documenting it or at
      least opened an issue for it? If so, please add a link to it.

-----

Fixes #2345

### CODE REVIEW

File: tests/func/test_run.py [hunk 1]
```diff
@@ -941,3 +942,14 @@ def test_bad_stage_fname(repo_dir, dvc_repo):
...
+
+
+def test_should_raise_on_stage_dependency(repo_dir, dvc_repo):
+    stage_path = dvc_repo.add(repo_dir.FOO)[0].path
+
+    with pytest.raises(StageFileAsOutputError):
+        dvc_repo.run(
+            cmd="python {} {} {}".format(repo_dir.CODE, repo_dir.FOO, "out"),
+            deps=[repo_dir.FOO, stage_path],
+            outs=["out"],
```
[the remark this answers is not in the document]
Comment (pared) [reply tests/func/test_run.py]: Makes sense

File: dvc/dependency/base.py [hunk 2]
```diff
@@ -15,11 +15,19 @@ def __init__(self, path):
         super(DependencyIsNotFileOrDirError, self).__init__(msg)
 
 
+class DependencyIsStageFileError(DvcException):
+    def __init__(self, path):
+        super(DependencyIsStageFileError, self).__init__(
+            "Stage file '{}' cannot be dependency.".format(path)
```
Comment (efiop) [inline dvc/dependency/base.py]:
```suggestion
            "Stage file '{}' cannot be a dependency.".format(path)
```

File: dvc/output/base.py [hunk 3]
```diff
@@ -32,6 +32,13 @@ def __init__(self, path):
         super(OutputAlreadyTrackedError, self).__init__(msg)
 
 
+class OutputIsStageFileError(DvcException):
+    def __init__(self, path):
+        super(OutputIsStageFileError, self).__init__(
+            "Stage file '{}' cannot be output.".format(path)
```
Comment (efiop) [inline dvc/output/base.py]:
```suggestion
            "Stage file '{}' cannot be an output.".format(path)
```

File: tests/func/test_run.py [hunk 4]
```diff
@@ -941,3 +943,25 @@ def test_bad_stage_fname(repo_dir, dvc_repo):
 
     # Check that command hasn't been run
     assert not os.path.exists("out")
+
+
+def test_should_raise_on_stage_dependency(repo_dir, dvc_repo):
+    stage_file_name = "name.dvc"
```
Comment (efiop) [inline tests/func/test_run.py]: No need to have a separate var for the string that is short than the name :) Just use "name.dvc" down below.

Comment (pared) [reply tests/func/test_run.py]: done

**Pull request merged**